Sources/GRPCCore/GRPCClient.swift

```diff
@@ -0,0 +1,823 @@
+/// set on the given transport.
+/// You can also use ``ClientInterceptor``s to implement cross-cutting logic which apply to all
+/// RPCs. Example uses of interceptors include authentication and logging.
+///
+/// ## Creating and configuring a client
+///
+/// The following example demonstrates how to create and configure a server.
+///
+/// ```swift
+/// // Create and add an in-process transport.
+/// let inProcessTransport = InProcessClientTransport()
+/// let client = GRPCClient(transport: inProcessTransport)
+///
+/// // Create and add some interceptors.
+/// client.interceptors.add(StatsRecordingServerInterceptors())
+///
+/// // Create and add some method configurations.
+/// let defaultConfiguration = MethodConfiguration(
+///     executionPolicy: ...,
+///     timeout: ...
+/// )
+/// let registry = MethodConfigurationRegistry(defaultConfiguration: defaultConfiguration)
+/// client.methodConfigurationOverrides = registry
+/// ```
+///
+/// ## Starting and stopping the client
+///
+/// Once you have configured the client, call ``run()`` to start it. Calling ``run()`` connects to the given
+/// transport.
+///
+/// ```swift
+/// // Start running the client.
+/// try await client.run()
+/// ```
+///
+/// The ``run()`` method won't return until the client has finished handling all requests. You can
+/// signal to the client that it should stop creating new request streams by calling ``close()``.
+/// This gives the client enough time to drain any requests already in flight. To stop the client more abruptly
+/// you can cancel the task running your client. If your application requires additional resources
+/// that need their lifecycles managed you should consider using [Swift Service
+/// Lifecycle](https://github.com/swift-server/swift-service-lifecycle).
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public final class GRPCClient: Sendable {
+  /// A collection of ``ClientInterceptor`` implementations which are applied to all accepted
+  /// RPCs.
+  ///
+  /// RPCs are intercepted in the order that interceptors are added. That is, a request sent from the client to
+  /// the server will first be intercepted by the first added interceptor followed by the second, and so on.
+  /// For responses from the server, they'll be applied in the opposite order.
+  public var interceptors: Interceptors {
+    get {
+      self.storage.withLockedValue { $0.interceptors }
+    }
+    set {
+      self.storage.withLockedValue { storage in
+        if case .notStarted = storage.state {
+          storage.interceptors = newValue
+        }
+      }
+    }
+  }
+
+  /// A ``MethodConfigurationRegistry`` containing ``MethodConfiguration``s for calls
+  /// made from this ``Client``.
+  ///
+  /// - Note: These configurations will override those configurations set in the ``ClientTransport``.
+  public var methodConfigurationOverrides: MethodConfigurationRegistry {
```

Comment:
These aren't just overrides, they're the default configuration too. I _think_ we probably need two of these: one for overrides one for defaults. 

I think we should stick these in a small wrapper type like the interceptors too:

```swift
client.methodConfiguration.overrides[foo] = ...
client.methodConfiguration.defaults[bar] = ...
```

I think this means that the storage for these doesn't need to offer a default value either, the client can just create the empty config if we didn't find one.

---

Naming here get's awkward: we'd have `GRPCClient.MethodConfiguration` as the wrapper containing the collections of what-used-to-be-called `ClientRPCExecutionConfiguration` for the `overrides`/`defaults`.

Maybe we should keep the existing name and nest the registry thing within `MethodConfiguration`?